### PR TITLE
[libc++] Fix the behavior of throwing `operator new` under -fno-exceptions

### LINKED ISSUE #60129: Operator new and -fno-exceptions
Current implementation of `operator new` in libcxx / libcxxabi effectively makes plain `new` and nothrow `new` to coincide when exceptions are disabled: both will return nullptr. However, there is one important difference: for nothrow `new` the constructor is not called when the return value is null. This does not happen for ordinary `new` in `-fno-exceptions` case making impossible to check the return value as constructor might just segfault. From the user perspective it might be better to stick to one of the following:
  * Just call `abort()`. This seems what is currently done for things like `std::__throw_out_of_range` in `-fno-except` case
  * Make `operator new` and nothrow `operator new` to behave identically and do not call constructor

Any thoughts on what might be the better solution?

### PR DESCRIPTION
**Pull request by ldionne**
In D144319, Clang tried to land a change that would cause some functions that are not supposed to return nullptr to optimize better. As reported in https://reviews.llvm.org/D144319#4203982, libc++ started seeing failures in its CI shortly after this change was landed.

As explained in D146379, the reason for these failures is that libc++'s throwing `operator new` can in fact return nullptr when compiled with exceptions disabled. However, this contradicts the Standard, which clearly says that the throwing version of `operator new(size_t)` should never return nullptr. This is actually a long standing issue. I've previously seen a case where LTO would optimize incorrectly based on the assumption that `operator new` doesn't return nullptr, an assumption that was violated in that case because libc++.dylib was compiled with -fno-exceptions.

Unfortunately, fixing this is kind of tricky. The Standard has a few requirements for the allocation functions, some of which are impossible to satisfy under -fno-exceptions:
1. `operator new(size_t)` must never return nullptr
2. `operator new(size_t, nothrow_t)` must call the throwing version and return nullptr on failure to allocate
3. We can't throw exceptions when compiled with -fno-exceptions

In the case where exceptions are enabled, things work nicely. `new(size_t)` throws and `new(size_t, nothrow_t)` uses a try-catch to return nullptr. However, when compiling the library with -fno-exceptions, we can't throw an exception from `new(size_t)`, and we can't catch anything from `new(size_t, nothrow_t)`. The only thing we can do from `new(size_t)` is actually abort the program, which does not make it possible for `new(size_t, nothrow_t)` to catch something and return nullptr.

This patch makes the following changes:
1. When compiled with -fno-exceptions, the throwing version of `operator new` will now abort on failure instead of returning nullptr on failure. This resolves the issue that the compiler could mis-compile based on the assumption that nullptr is never returned. This constitutes an API and ABI breaking change for folks compiling the library with -fno-exceptions (which is not the general public, who merely uses libc++ headers but use a shared library that has already been compiled). This should mostly impact vendors and other folks who compile libc++.dylib themselves.

2. When the library is compiled with -fexceptions, the nothrow version of `operator new` has no change. When the library is compiled with -fno-exceptions, the nothrow version of `operator new` will now check whether the throwing version of `operator new` has been overridden. If it has not been overridden, then it will use an implementation equivalent to that of the throwing `operator new`, except it will return nullptr on failure to allocate (instead of terminating). However, if the throwing `operator new` has been overridden, it is now an error NOT to also override the nothrow `operator new`. Indeed, there is no way for us to implement a valid nothrow `operator new` without knowing the exact implementation of the throwing version.

In summary, this change will impact people who fall into the following intersection of conditions:
- They use the libc++ shared/static library built with `-fno-exceptions`
- They do not override `operator new(..., std::nothrow_t)`
- They override `operator new(...)` (the throwing version)
- They use `operator new(..., std::nothrow_t)`

We believe this represents a small number of people.

Fixes #60129
rdar://103958777

Differential Revision: https://reviews.llvm.org/D150610

### CODE REVIEW

**Comment by ldionne** [comment]
This change is extremely tricky but also pretty necessary -- pinging @llvm/libcxx-vendors .

**If you are a vendor and you build the libc++ shared or static library with `-fno-exceptions`, you must take the time to read the commit message + release note and understand the implications of this change.**

This change does not affect users merely building with ``-fno-exceptions`` if `libc++.dylib`/`libc++.a` has been built with exceptions enabled. But it affects anyone who's building `libc++.dylib`/`libc++.a` with exceptions disabled.

The alternative to this patch is to change the Standard to allow `operator new` to return `nullptr` when exceptions are disabled, which is not something that's acknowledged by the Standard. And then this patch would instead turn into a Clang patch that makes it stop assuming that `operator new` doesn't return `nullptr`.

File: libcxx/src/new.cpp [hunk 1]
```diff
@@ -38,39 +44,53 @@ static void* operator_new_impl(std::size_t size) noexcept {
...
   if (p == nullptr)
-    throw std::bad_alloc();
-#  endif
+    __throw_bad_alloc_shim();
   return p;
 }
 
 _LIBCPP_WEAK void* operator new(size_t size, const std::nothrow_t&) noexcept {
+#  ifdef _LIBCPP_HAS_NO_EXCEPTIONS
+  _LIBCPP_ASSERT_SHIM(
```
Comment (ldionne) [inline libcxx/src/new.cpp:54]: @var-const This is a rare example of a case where I think it makes sense to unconditionally have a `_LIBCPP_ASSERT` -- this should be enabled regardless of the hardening mode IMO. It is similar to when we call `std::__throw_logic_error()` and friends with `-fno-exceptions` -- that aborts the program regardless of the hardening mode.

**Review by rnk — COMMENTED** [review]
+@jyknight for new + EH thoughts

File: libcxx/include/new [hunk 2]
```diff
@@ -214,15 +215,15 @@ inline constexpr destroying_delete_t destroying_delete{};
 
 #if !defined(_LIBCPP_ABI_VCRUNTIME)
 
-_LIBCPP_NODISCARD_AFTER_CXX17 _LIBCPP_OVERRIDABLE_FUNC_VIS void* operator new(std::size_t __sz) _THROW_BAD_ALLOC;
+_LIBCPP_NODISCARD_AFTER_CXX17 _LIBCPP_OVERRIDABLE_FUNC_VIS _LIBCPP_MAKE_OVERRIDABLE_FUNCTION_DETECTABLE void* operator new(std::size_t __sz) _THROW_BAD_ALLOC;
```
[the remark this answers is not in the document]
Comment (rnk) [reply libcxx/include/new:218]: Example of setting the section on the definition but not the declaration, has no warnings: https://gcc.godbolt.org/z/5Ed7vGxfE

[the remark this answers is not in the document]
Comment (ldionne) [reply libcxx/include/new:218]: Yeah, looks like that should work!

File: libcxx/include/__overridable_function [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#ifndef _LIBCPP___OVERRIDABLE_FUNCTION
+#define _LIBCPP___OVERRIDABLE_FUNCTION
+
+#include <__config>
+#include <cstdint>
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#define _LIBCPP_MAKE_OVERRIDABLE_FUNCTION_DETECTABLE                                                                   \
+  __attribute__((__section__("__TEXT,__lcxx_override,regular,pure_instructions")))
```
[the remark this answers is not in the document]
Comment (ldionne) [reply libcxx/include/__overridable_function:19]: Yeah, that's right. In a way, this is "fortunate" since this whole `__overridable_function` thing is only used to power the assertion we have in case a user forgot to override the `nothrow` version. So we could just disable it on some platforms -- but in practice it would be really nice to have this diagnostic on all supported platforms, since this is so tricky.

[the remark this answers is not in the document]
Comment (rnk) [reply libcxx/include/__overridable_function:19]: To port this to ELF, I think you could use `__attribute__((section("lcxx_override")))`, and then declare external `__start_lcxx_override` and `__stop_lcxx_override` symbols. That should work with all the linkers we care about. COFF can probably be done, but I'm sure it's more involved and probably not worth doing.

**Review by rnk — COMMENTED** [review]
These are just thoughts and suggestions to be taken or left. Other reviewers should please feel free to approve.

File: libcxx/src/new.cpp (hunk 1 above)
Comment (EricWF) [reply libcxx/src/new.cpp:54]: I think you're forgetting how hot of a path this is. I don't think we want to be doing anything more than we need in the majority of cases.

**Review by EricWF — COMMENTED** [review]
Are there tests that we actually correctly call the overridden throwing version from the overridden non-throwing version? 

Also, I wonder how much we really care about the library being built without exceptions vs allowing the library to be used without exceptions.

**Comment by ldionne** [comment]
> Are there tests that we actually correctly call the overridden throwing version from the overridden non-throwing version?

Yes, see for example `libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.single/new.size_nothrow.replace.indirect.pass.cpp`.

> Also, I wonder how much we really care about the library being built without exceptions vs allowing the library to be used without exceptions.

We currently support both ways of building the library (`-fexceptions` and `-fno-exceptions`). I don't know how widely used the `-fno-exceptions` variant is used and we could discuss dropping it (and forcing people to always build the library itself with `-fexceptions`), but IMO this should be a separate change. In practice, I suspect that a lot of folks in the embedded world might be building the library with `-fno-exceptions`.

File: libcxx/src/new.cpp [hunk 4]
```diff
@@ -7,6 +7,8 @@
 //===----------------------------------------------------------------------===//
 
 #include <__memory/aligned_alloc.h>
+#include <__overridable_function>
```
[the remark this answers is not in the document]
Comment (ldionne) [reply libcxx/src/new.cpp:10]: So this file needs to be shared between libc++abi and libc++, unfortunately, so it's a bit easier to throw in under `include`.

File: libcxx/src/new.cpp (hunk 1 above)
Comment (ldionne) [reply libcxx/src/new.cpp:54]: We could make this part of the `DEBUG_LITE` mode then, I guess.

**Comment by EricWF** [comment]
> We currently support both ways of building the library (-fexceptions and -fno-exceptions). I don't know how widely used the -fno-exceptions variant is used and we could discuss dropping it (and forcing people to always build the library itself with -fexceptions), but IMO this should be a separate change. In practice, I suspect that a lot of folks in the embedded world might be building the library with -fno-exceptions.

My only concern is that this change introduces a bunch of black-magic, which I suspect is pretty fragile (but I don't know), and I would like to avoid that if possible.

**Comment by ldionne** [comment]
> I'm tempted to just do the non-conforming thing here (because `-fno-exceptions` is non-conforming), and just assume that when you're building with `-fno-exceptions`, you're not trying to replace the throwing operator new, and you'll be fine if we have the throwing kind call the non-throwing kind.

So IIUC you are suggesting the following behavior under `-fno-exceptions`?

```
operator new(size_t) // We assume it is never overridden, we call new(size_t, nothrow_t) and we abort if that fails
operator new(size_t, nothrow_t) // Users may override it. Default implementation tries to allocate memory using malloc
```

Under `-fexceptions`, the behavior would stay as it is right now, i.e.

```
operator new(size_t) // Can be overridden, by default tries to allocate with malloc and throws if that fails
operator new(size_t, nothrow_t) // Users may override, by default it uses new(size_t) and uses try-catch to return nullptr on failure
```

Is that your suggestion? The problem I see with this is that we end up with inverted dependencies between `new(size_t)` and `new(size_t, nothrow_t)` depending on whether exceptions are enabled. With `-fexceptions`, we get the usual dependency `new(size_t, nothrow_t) calls new(size_t)`, but with `-fno-exceptions` we have `new(size_t) calls new(size_t, nothrow_t)`. IMO this change in semantics is an issue, especially considering that users can use a different exception mode than what the dylib is built with. That seems like a recipe for confusion.

Another issue is that concretely, 99% of people replace `new(size_t)` and probably never bother to replace (or even use) `new(size_t, nothrow_t)`. So I feel like expecting users to override `new(size_t, nothrow_t)` instead of `new(size_t)` under `-fno-exceptions` is probably unrealistic.

The currently-proposed approach kinda sucks too (because we don't know how to implement `new(size_t, nothrow_t)` if you replace `new(size_t)`), but at least it's consistent in every other regard (AFAICT) and it is as Standards conforming as possible.

If we could design this from scratch, what would we do? I think we would likely specify that `new(size_t)` is implemented as calling `new(size_t, nothrow_t)` and doing `if (result == nullptr) throw std::bad_alloc();`. Then the common way to override `new` amongst users would be to override `new(size_t, nothrow_t)` instead of `new(size_t)` and they would immediately get the right behavior. Under `-fno-exceptions`, we would simply `abort()` inside `new(size_t)` in case the `nothrow_t` version returned `nullptr`. Do we think there is even a slim chance of ever changing that in the Standard?

**Comment by jyknight** [comment]
I think we can check whether it's been overridden more easily using an asm alias like this (when libc++ is built under -fno-exceptions):
```
#include <cstdlib>
#include <new>

__attribute__((weak)) void* operator new(std::size_t size) {
  void* mem = malloc(size);
  if (!mem) abort();
  return mem;
}

// TODO: ifdef for MSVC mangling.
static void* original_operator_new(std::size_t size) __attribute__((alias(__USER_LABEL_PREFIX__ "_Znwm")));

__attribute__((weak)) void* operator new(std::size_t size, std::nothrow_t) noexcept {
  if (static_cast<void*(*)(std::size_t)>(&::operator new) != original_operator_new)
    abort();
  return malloc(size);
}
```

**Comment by EricWF** [comment]
@ldionne  You've fully convinced me on this approach. Thank you for the explanation.

File: libcxx/include/__overridable_function (hunk 3 above)
[the remark this answers is not in the document]
Comment (ldionne) [reply libcxx/include/__overridable_function:23]: I think I finally got this to work for ELF! The next revision will tell.

**Comment by ldionne** [comment]
> I think we can check whether it's been overridden more easily using an asm alias like this (when libc++ is built under -fno-exceptions):

This doesn't work on Darwin (apparently we don't support the alias attribute on that platform). I could investigate doing this all non-Darwin platforms, however that would make the utility a lot less generic than in my latest patch, since I'd have to handle each function I want to check for overrides individually.

File: libcxx/src/new.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (philnik777) [reply libcxx/src/new.cpp:10]: We already have `#include "include/refstring.h" // from libc++` inside `libcxxabi/src/stdlib_stdexcept.cpp`, so I don't think that should be a blocker from moving it to `src/`.

**Review by philnik777 — COMMENTED** [review]
LGTM % nit.

**Review by philnik777 — APPROVED** [review]

**Comment by DianQK** [comment]
Thanks for the fix! I'll merge the blocked patches in a later two days.
